[rpc] fix getchaintxstats()

AkioNak opened the pull request:
1. calculate nblocks more adaptive.
  -> set default nblocks to min (blocks for 1 month, target block's height - 1)
  -> before PR: if not specify nblocks-parameter, illegal parameter error will happen when target block height is below nblocks.
2. correct error message.
  -> nblocks accepts [1 .. block's height -1] . so add a word "-1".
3. add check 0-divide.
  -> if nTimeDiff = 0 then use UniValue(UniValue::VNULL) and returns {... "txrate": null} .
  -> before PR: if nTimeDiff = 0 then returns {... "txrate":} and bitcoin-cli cannot handle the response.

AkioNak committed:
Fix getchaintxstats()

1. Calculate nblocks more adaptive.
   If not specify nblocks-parameter, illegal parameter error
   will happen when target block height is below blocks for 1 month.
   To avoid this error, set default nblocks to
   min(blocks for 1 month, target block's height - 1)
   And allowing 0 so that this RPC works good even if target block is
   genesis block or 1st block.
2. Correct error message.
   nblocks accepts [0 .. block's height -1] . so fix as following:
   "Invalid block count: should be between 0 and the block's height - 1"
3. Add check 0-divide.
   If nTimeDiff = 0 then returns {... "txrate":} and
   bitcoin-cli cannot handle the response.
   To avoid this error, do not return "txrate" if nTimeDiff = 0.
4. Add following 3 elements to the return object.
   1) 'window_block_count' : Size of the window in number of blocks.
   2) 'window_tx_count' : The number of transactions in the window.
   3) 'window_interval' : The elapsed time in the window.
   They clarify how 'txrate' is calculated. 2) and 3) are returned
   only if 'window_block_count' is a positive value.
5. Improve help text for 'time' as following.
   'The timestamp for the final block in the window in UNIX format.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1516,8 +1512,14 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
         }
     }
 
+    blockcount = std::min(blockcount, pindex->nHeight - 1);
```
Edge case: chain has 1 block only - `blockcount` will be 0 - should be a different error?

AkioNak replied:
@promag Thank you for you review.
I added a commit that throws a new RPC_INVALID_PARAMETER if block's height below 2.

promag replied:
I wonder if the error could be avoided by making the algorithm compatible with 1 block chain.

promag reviewed (commented):
utACK e73f8bc.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1512,6 +1512,9 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
     }
 
+    if (pindex->nHeight < 2) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Block's height should be 2 or more.");
```
IMO the error should be `RPC_MISC_ERROR` since it's not a parameter. The message should be something like `Chain height should be greater than 1`.

promag replied:
I wonder if the error could be avoided by making the algorithm compatible with 1 block chain.

AkioNak replied:
I think tx-rate for 1st block (and genesis block) can not calculate (or meaningless).

> be avoided by making the algorithm compatible with 1 block chain.
So, I think we can treat this situation equal to nTimeDiff = 0.

promag commented:
Can you improve the test in `bitcoin/test/functional/blockchain.py`?

AkioNak commented:
Yes, I will add some tests in `bitcoin/test/functional/blockchain.py`.

AkioNak committed:
Add some tests for getchaintxstats

1. Add a test for no parameters.
2. Add a test for the block's height = 1.
3. Add a test for nblocks is out of range.

AkioNak commented:
@promag I avoided the error about low blockheight and added some tests.

promag reviewed (commented):
IMO `throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid block count: should be between 1 and the block's height - 1")` should be only for `< 1`case. `nblocks` is the size of the window, not really necessary to satisfy. Even the default value will throw for a chain with less than a month.

Also, `CBlockIndex::GetAncestor` return `nullptr` for out of bounds, just check that result.

AkioNak commented:
@promag 
> IMO throw `JSONRPCError(RPC_INVALID_PARAMETER, "Invalid block count: should be between 1 and the block's height - 1")` should be only for < 1case. 

When blockheight = 1, nblocks = 1 does not satisfy this condition. (blockheight> = 2 is required)
So, same as your previous  suggestion, I think it would be better to throw different error depending on whether the block(hash) is specified by the parameter or not.

>Also, `CBlockIndex::GetAncestor` return `nullptr` for out of bounds, just check that result.

Of course I can add to check it , but I think that the value of the parameter is always within the range. Am I missing something?

jnewbery commented on test/functional/blockchain.py:
```diff
@@ -56,6 +56,22 @@ def _test_getchaintxstats(self):
 
+        b1 = self.nodes[0].getblock(self.nodes[0].getblockhash(1))
+        b200 = self.nodes[0].getblock(self.nodes[0].getblockhash(200))
+        ntimediff = b200['mediantime'] - b1['mediantime']
```
nit: name the variable `time_diff`. (the `n` prefix in the C++ code is no longer the style we use, and shouldn't be carried over to the python code)

jnewbery reviewed (commented):
Thanks for the PR @AkioNak . Looks great. A couple of suggestions:

- perhaps add a `"block_count"` element to the return object, so it's obvious over how many blocks the `"txrate"` is being calculated.
- I'd prefer to not return a `txrate` at all if `nTimeDiff` is zero, ie:

```
if (nTimeDiff > 0) {
    ret.push_back(Pair("txrate", ((double)nTxDiff) / nTimeDiff);
}
```

That's just personal preference though. Tested ACK even without those changes.

One small style nit inline.

jnewbery commented:
I also recommend that you update the commit messages to be a bit more meaningful. I don't think this project has commit message style rules, but there are some general tips here: https://chris.beams.io/posts/git-commit/

AkioNak commented:
@jnewbery Thank you for your suggestion. I agree with them.

promag commented:
BTW, if `time_diff` is zero then `pindexPast` and `pindex` are the same?

AkioNak commented:
@promag No. Even if `time_diff` is zero, `pindexPast` and `pindex` may be different.
(For example, `generates 100` on regtest)
Because `GetMedianTimePast()` returns mediantime of last 11 block's time (except the block itself),

promag commented:
IMO the response could have both values `txratecount` and `txrateinterval`.

AkioNak commented:
@promag  In the edge case (block height <2), since there is no comparison target, its values are both zero. If that is OK then I will add it.

AkioNak commented:
@jnewbery I have just done followings.

- Add `"block_count"` element to return object.
- Not to return `"txrate"` if `nTimeDiff` is zero.
- Rename variable `time_diff`.
- Fix commit comment.

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -1480,6 +1480,7 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
             "{\n"
             "  \"time\": xxxxx,        (numeric) The timestamp for the statistics in UNIX format.\n"
             "  \"txcount\": xxxxx,     (numeric) The total number of transactions in the chain up to that point.\n"
+            "  \"block_count\": xxxxx, (numeric) The block height that ends the window.\n"
```
This seems to be a confusing name; I would associate `block_count` with the size of the window over which `txrate` is computed. To denote the block height that ends the window, I would use `height`.

jnewbery replied:
I agree. My earlier suggestion was for `block_count` to be the size of the window.

AkioNak replied:
@sipa Thank you for your review. I made a mistake.
@jnewbery I misunderstood. Same as your suggestion, I think the size of the window is better than block height as result of this command.

AkioNak commented:
@sipa @jnewbery I fixed what `"block_count"` indicates and its description.

AkioNak commented:
@promag I added  both `txratecount` and `txrateinterval` to the return object.

AkioNak commented:
I resolved the conflicts.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1479,6 +1479,9 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
             "  \"txcount\": xxxxx,     (numeric) The total number of transactions in the chain up to that point.\n"
+            "  \"block_count\": xxxxx, (numeric) Size of the window in number of blocks.\n"
+            "  \"txratecount\": xxxxx, (numeric) The number of transactions in the window.\n"
+            "  \"txrateinterval\": xxxxx, (numeric) The elapsed time in the window.\n"
```
I don't love the `txratecount` and `txrateinterval` names (rate and count are fundamentally different things, and `txrateinterval` mentions tx but not time, but the value returned is time).

I suggest:
- `window_block_count`
- `window_tx_count`
- `window_interval`

Note that new RPC arguments and return values should be snake case.

jnewbery replied:
Also perhaps add a comment to the help text that these values are only returned if `block_count` is > 0

AkioNak replied:
I agree to rename those names and to add condition to the help text.
But I wonder `txcount`/`txrate` should rename to `tx_count`/`tx_rate` or remain.
If the name of txcount is remained , I like `window_txcount`  than `window_tx_count`.
Which do you think is better?

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1515,8 +1514,18 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+    if (!request.params[0].isNull()) {
+        blockcount = request.params[0].get_int();
+
+        if (blockcount < 1 || blockcount >= pindex->nHeight) {
```
Any reason to disallow 0? `blockcount` can be 0 if we run this on a node with just the genesis block, so why not allow it to be set to 0 when running with a longer chain?

AkioNak replied:
I think the reason to disallow 0 is that this RPC was originally intended to be able to calculate `txrate`.
Therefore if user specifies 0 by parameter, PRC should returns illegal parameter error.
So, for only genesis block and 1st block, I think it can allow 0 internally when user is not specify block count in 1st parameter, and can avoid error to be occured.

But now, it is easier to understand if we allow 0.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1526,7 +1535,14 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+        ret.push_back(Pair("txratecount", nTxDiff));
+        ret.push_back(Pair("txrateinterval", nTimeDiff));
+    }
+    if (nTimeDiff > 0) {
```
Perhaps move this up into the `if (blockcount > 0)` code block (since `nTimeDiff` can only be non-zero when `blockcount` > 0)? Up to you.

AkioNak replied:
I agree to move it.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1479,6 +1479,9 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
             "{\n"
             "  \"time\": xxxxx,        (numeric) The timestamp for the statistics in UNIX format.\n"
```
Suggestion: improve this help message as part of this PR. Perhaps:

`The timestamp for the final block in the window in UNIX format.`

AkioNak replied:
I agree to improve that help message.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1526,7 +1535,14 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
     ret.push_back(Pair("txcount", (int64_t)pindex->nChainTx));
-    ret.push_back(Pair("txrate", ((double)nTxDiff) / nTimeDiff));
+    if (blockcount > 0) {
+        ret.push_back(Pair("block_count", blockcount));
```
Perhaps display `block_count` if it is equal to zero? (but not the other two values since they're meaningless if block_count == 0

AkioNak replied:
I agree to display `block_count` even if it is equal to zero.

jnewbery reviewed (commented):
Looks really good now. I have a few nits inline.

You'll need to squash your commits down. Use `git rebase -i` and feel free to message me on IRC if you want some pointers on how to do that.

AkioNak commented:
@jnewbery Thank you for your various suggestion. I accepted with them and squashed commits for file by file.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1477,9 +1477,12 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+            "  \"time\": xxxxx,                (numeric) The timestamp for the final block in the window in UNIX format.\n"
+            "  \"txcount\": xxxxx,             (numeric) The total number of transactions in the chain up to that point.\n"
+            "  \"window_block_count\": xxxxx,  (numeric) Size of the window in number of blocks.\n"
+            "  \"window_tx_count\": xxxxx,     (numeric) The number of transactions in the window. Only returned if \"window_block_count\" is > 0.\n"
```
Nit, above it's `txcount` so here should be `window_txcount`?

jnewbery replied:
I think this is fine. New arguments and return values should use snake_case.

AkioNak replied:
I agree to @jnewbery ,too.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1477,9 +1477,12 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+            "  \"txcount\": xxxxx,             (numeric) The total number of transactions in the chain up to that point.\n"
+            "  \"window_block_count\": xxxxx,  (numeric) Size of the window in number of blocks.\n"
+            "  \"window_tx_count\": xxxxx,     (numeric) The number of transactions in the window. Only returned if \"window_block_count\" is > 0.\n"
+            "  \"window_interval\": xxxxx,     (numeric) The elapsed time in the window. Only returned if \"window_block_count\" is > 0.\n"
```
Include unit (seconds)?

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1515,8 +1514,18 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+    if (!request.params[0].isNull()) {
+        blockcount = request.params[0].get_int();
+
+        if (blockcount < 0 || (blockcount > 0 && blockcount >= pindex->nHeight)) {
```
Remove `blockcount > 0`?

kallewoof replied:
`blockcount == 0` is not an error so `blockcount > 0` has to remain, I believe.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1515,8 +1514,18 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
 
-    if (blockcount < 1 || blockcount >= pindex->nHeight) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid block count: should be between 1 and the block's height");
+    if (pindex->nHeight < 1) {
```
Move to `else` like:
```cpp
if (!request.params[0].isNull()) {
    ...
} else {
    blockcount = std::max(0, std::min(blockcount, pindex->nHeight - 1));
}
```

jnewbery replied:
I agree that this is clearer

AkioNak replied:
@promag Thanks for a good suggestion.
@promag @jnewbery Is it better to reverse the conditions?
```c++
if (request.params[0].isNull()) {
    blockcount = std::max(0, std::min(blockcount, pindex->nHeight - 1));
} else {
    ...
}
```

jnewbery replied:
No preference either way!

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1515,8 +1514,18 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+    if (!request.params[0].isNull()) {
+        blockcount = request.params[0].get_int();
+
+        if (blockcount < 0 || (blockcount > 0 && blockcount >= pindex->nHeight)) {
```
Agree

AkioNak replied:
@promag @jnewbery I think that this condition is necessary if pindex points to `genesis block`.

jnewbery commented:
@AkioNak thanks for sticking with this. Looks really good now. Just a few more nits.

AkioNak commented:
@promag @jnewbery @kallewoof I fixed followings.
1. add unit(seconds) to the help text of `window_interval`.
2. calculate `blockcount` using `std::max()` rather than if-clause.

jnewbery commented:
Looks good to me! tested ACK 07704c1b3768d6c290046c783063644fc7b7d1da.

TheBlueMatt commented:
Postumous utACK non-test parts of 07704c1b3768d6c290046c783063644fc7b7d1da
